### PR TITLE
Attach OIDC to Iam role

### PR DESCRIPTION
**Pull request by NishanthSpShetty**
# Description

Provide cluster components, the access to aws resources.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# why

- Allow service running inside the cluster to access aws resources.

# How Has This Been Tested?
 TODO

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: proto/netbookai/spawner/spawner.proto [hunk 1]
```diff
@@ -77,7 +77,11 @@ service SpawnerService {
   rpc GetContainerRegistryAuth(GetContainerRegistryAuthRequest)
       returns (GetContainerRegistryAuthResponse) {}
 
-rpc CreateContainerRegistryRepo(CreateContainerRegistryRepoRequest) returns (CreateContainerRegistryRepoResponse) {}
+  rpc CreateContainerRegistryRepo(CreateContainerRegistryRepoRequest)
+      returns (CreateContainerRegistryRepoResponse) {}
+
+  rpc ConnectClusterOIDCToTrustPolicy(ConnectClusterOIDCToTrustPolicyRequest)
```
Comment (sid255) [inline proto/netbookai/spawner/spawner.proto:83]: Can we name this something generic like `RegisterClusterOIDC` ?

File: pkg/service/aws/oidc.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	if *cluster.Status != eks.ClusterStatusActive {
+		a.logger.Info(ctx, "cluster is not active yet", "status", *cluster.Status)
+		return nil, fmt.Errorf("cluster is not active, curent state: %s", *cluster.Status)
+	}
+	if cluster.Identity == nil {
+		a.logger.Info(ctx, "cluster doesnt have identity", "identity", nil)
+		return nil, errors.New("cluster identity is nil")
+	}
+
+	issuer := *cluster.Identity.Oidc.Issuer
```
Comment (sid255) [inline pkg/service/aws/oidc.go:98]: Can we check if Oidc is non-nil and so is the issuer? To be on safer side?

File: pkg/service/aws/oidc.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+	//check if we already created a open id oidcProvider
+	_, err = iamClient.GetOpenIDConnectProviderWithContext(ctx, &iam.GetOpenIDConnectProviderInput{
+		OpenIDConnectProviderArn: &providerArn,
+	})
+
+	if err != nil {
+		a.logger.Error(ctx, "failed to fetch open id provider", "error", err)
+
+		//TODO: where is this coming from ? @mani
+		thumbprint := "9e99a48a9960b14926bb7f3b02e22da2b0ab7280"
```
Comment (sid255) [inline pkg/service/aws/oidc.go:120]: Replicate functionality from https://github.com/gruntwork-io/kubergrunt/blob/master/eks/oidc.go to get thumbprint.

File: pkg/service/aws/oidc.go [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+	})
+	if err != nil {
+		a.logger.Error(ctx, "failed to get the oidc role", "error", err)
+		return nil, errors.Wrap(err, "failed to get the role")
+	}
+
+	//get the current policy doc and append current cluster statement
+	newPolicy, err := generateTrustPolicyDocument(*role.Role.AssumeRolePolicyDocument, providerArn, clusterUrl)
+	if err != nil {
+		a.logger.Error(ctx, "generateTrustPolicyDocument ")
```
Comment (sid255) [inline pkg/service/aws/oidc.go:144]: Log the  error

File: pkg/service/aws/oidc.go [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+// oidc endpoint
+const trustPolicTemplate = `{
+			"Effect": "Allow",
+			"Principal": {
+				"Federated": "%s"
+			},
+			"Action": "sts:AssumeRoleWithWebIdentity",
+			"Condition": {
+				"StringEquals": {
+					"%s:sub": "system:serviceaccount:oidc:awskube2iam",
```
Comment (sid255) [inline pkg/service/aws/oidc.go:29]: @FiReBeNdEr07 can you confirm the service account namespace and name?
@NishanthSpShetty can we pick the namespace `oidc` and service account name `awskube2iam` from constants?

**Review by sid255 — CHANGES_REQUESTED** [review]

File: pkg/service/aws/oidc.go (hunk 2 above)
Comment (NishanthSpShetty) [reply pkg/service/aws/oidc.go:98]: yeah, have added.

File: pkg/service/aws/oidc.go (hunk 4 above)
Comment (NishanthSpShetty) [reply pkg/service/aws/oidc.go:144]: updated.

File: pkg/service/aws/oidc.go [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+		a.logger.Info(ctx, "cluster doesnt have identity", "identity", nil)
+		return nil, errors.New("cluster identity is nil")
+	}
+
+	if cluster.Identity.Oidc == nil {
+		a.logger.Info(ctx, "cluster doesnt have oidc identity", "identity.oidc", nil)
+		return nil, errors.New("cluster oidc identity is nil")
+	}
+
+	issuer := *cluster.Identity.Oidc.Issuer
```
Comment (sid255) [inline pkg/service/aws/oidc.go:206]: Can we check issuer is nonempty?

File: pkg/service/aws/oidc.go (hunk 5 above)
Comment (NishanthSpShetty) [reply pkg/service/aws/oidc.go]: updated.

File: pkg/service/aws/oidc.go (hunk 6 above)
Comment (sid255) [reply pkg/service/aws/oidc.go:217]: This is not handled

Comment (NishanthSpShetty) [reply pkg/service/aws/oidc.go:217]: had added nil check issuer is a pointer, so if its not there it should be nil, not empty. To be on safer side, will add empty check too

**Review by sid255 — APPROVED** [review]
LGTM

File: pkg/service/aws/oidc.go [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+	"github.com/pkg/errors"
+	"gitlab.com/netbook-devs/spawner-service/pkg/config"
+	"gitlab.com/netbook-devs/spawner-service/pkg/service/constants"
+	proto "gitlab.com/netbook-devs/spawner-service/proto/netbookai/spawner"
+)
+
+//arn:aws:iam::965734315247:oidc-provider
+//
+// oidc endpoint
+const trustPolicTemplate = `{
```
Comment (NishanthSpShetty) [inline pkg/service/aws/oidc.go:28]: @FiReBeNdEr07

**Pull request merged**